example/src/main/scala/tamer/example/Main.scala

```diff
@@ -1,41 +1,48 @@
 package tamer
 package example
 
-import java.time.Instant
-import java.time.temporal.ChronoUnit._
-
+import tamer.config.Config
+import tamer.db.Db.Datable
+import tamer.db.{ConfigDb, DbTransactor, TamerDBConfig}
+import tamer.kafka.Kafka
+import zio.blocking.Blocking
+import zio._
 import doobie.implicits.legacy.instant._
 import doobie.syntax.string._
-import zio.UIO
+import log.effect.LogWriter
+import log.effect.zio.ZioLogWriter.log4sFromName
+import tamer.db.ConfigDb.{DbConfig, QueryConfig}
+
+import java.time.temporal.ChronoUnit._
+import java.time.{Duration, Instant}
 
-final case class State(from: Instant, to: Instant)
 final case class Key(id: String)
-final case class Value(id: String, name: String, description: Option[String], modifiedAt: Instant)
+final case class Value(id: String, name: String, description: Option[String], modifiedAt: Instant) extends Datable(modifiedAt)
+object Value {
+  implicit val ordering: Ordering[Value] = (x: Value, y: Value) => x.modifiedAt.compareTo(y.modifiedAt)
+}
 
-object Source {
-  private[this] implicit final class InstantOps(private val instant: Instant) extends AnyVal {
-    def plus5Minutes: Instant = instant.plus(5, MINUTES)
-    def minus60Days: Instant  = instant.minus(60, DAYS)
-    def orNow: UIO[Instant] =
-      UIO(Instant.now).map {
-        case now if instant.isAfter(now) => now
-        case _                           => instant
-      }
-  }
-  final val setup = UIO(Instant.now.truncatedTo(DAYS)).map { bootTime =>
-    val sixtyDaysAgo = bootTime.minus60Days
-    Setup.avro(
-      State(sixtyDaysAgo, sixtyDaysAgo.plus5Minutes)
-    )(s => sql"""SELECT id, name, description, modified_at FROM users WHERE modified_at > ${s.from} AND modified_at <= ${s.to}""".query[Value])(
-      v => Key(v.id),
-      s => {
-        case QueryResult(_, Nil) => s.to.plus5Minutes.orNow.map(State(s.from, _))
-        case QueryResult(_, values) =>
-          val max = values.map(_.modifiedAt).max // if we can't order in the query we need to do it here...
-          max.plus5Minutes.orNow.map(State(max, _))
-      }
+object Main extends zio.App {
```

Comment:
Feels like we could still be composing this more minimally here in the example. There should be a way to provide all the below in the core module, bar the db parts.

Reply:
Yes, I didn't spend much time on this and I'll leave tidying up the layers last or even in another PR. I expect it to be rather mechanical.

Reply:
also will inline `keyExtract`